lib/internal/worker/messaging.js

```diff
@@ -0,0 +1,229 @@
+// We need one for the operation status (performing / performed) and one for the result (success / failure).
+const WORKER_MESSAGING_SHARED_DATA = 2 * 4;
+const WORKER_MESSAGING_STATUS_INDEX = 0;
+const WORKER_MESSAGING_RESULT_INDEX = 1;
+
+// This event handler is always executed on the main thread only
+function handleMessageFromThread(message) {
+  switch (message.type) {
+    case messageTypes.REGISTER_MAIN_THREAD_PORT:
+      {
+        const { thread, port } = message;
+
+        // Register the port
+        threadsPorts.set(thread, port);
+
+        // Handle messages on this port
+        // When a new thread wants to register a children
+        // this take care of doing that.
+        // This way any thread can be linked to the main one.
+        port.on('message', handleMessageFromThread);
+
+        // Never block the thread on this port
+        port.unref();
+      }
+
+      break;
+    case messageTypes.UNREGISTER_MAIN_THREAD_PORT:
+      threadsPorts.get(message.thread).close();
+      threadsPorts.delete(message.thread);
+      break;
+    case messageTypes.SEND_MESSAGE_TO_WORKER:
+      {
+        // Send the message to the target thread
+        const { source, destination, value, transferList, memory } = message;
+        sendMessageToWorker(source, destination, value, transferList, memory);
+      }
+      break;
+  }
+}
+
+function handleMessageFromMainThread(message) {
+  switch (message.type) {
+    case messageTypes.RECEIVE_MESSAGE_FROM_WORKER:
+      receiveMessageFromWorker(message.source, message.value, message.memory);
+      break;
+  }
+}
+
+function sendMessageToWorker(source, destination, value, transferList, memory) {
+  // We are on the main thread, we can directly process the message
+  if (destination === 0) {
+    receiveMessageFromWorker(source, value, memory);
+    return;
+  }
+
+  // Search the port to the target thread
+  const port = threadsPorts.get(destination);
+
+  if (!port) {
+    const status = new Int32Array(memory);
+    AtomicsStore(status, WORKER_MESSAGING_RESULT_INDEX, 0);
+    AtomicsStore(status, WORKER_MESSAGING_STATUS_INDEX, 1);
+    AtomicsNotify(status, WORKER_MESSAGING_STATUS_INDEX, 1);
+    return;
+  }
+
+  port.postMessage(
+    {
+      type: messageTypes.RECEIVE_MESSAGE_FROM_WORKER,
+      source,
+      destination,
+      value,
+      memory,
+    },
+    transferList,
+  );
+}
+
+function receiveMessageFromWorker(source, value, memory) {
+  const status = new Int32Array(memory);
+  const deliverable = Number(process.listenerCount('workerMessage') > 0);
+
+  // If somebody is listening, emit the message
+  if (deliverable) {
+    process.emit('workerMessage', value, source);
+  }
+
+  // Populate the result
+  AtomicsStore(status, WORKER_MESSAGING_RESULT_INDEX, deliverable);
+  AtomicsStore(status, WORKER_MESSAGING_STATUS_INDEX, 1);
+  AtomicsNotify(status, WORKER_MESSAGING_STATUS_INDEX, 1);
+}
+
+function createMainThreadPort(thread) {
```

Comment:
```suggestion
function createMainThreadPort(threadId) {
```
I find it easier to read if `thread` is change to `threadId` here and at several other places in this file. for me `thread` reads similar as `worker` so an object and not just the numeric Id.
Feel free to ignore if you don't agree.

Reply:
I actually agree, thanks. Changed.